libmamba/include/mamba/util/graph.hpp

```diff
@@ -118,19 +96,48 @@ namespace mamba::util
         std::size_t m_number_of_edges = 0;
     };
 
-    template <typename Node, typename Derived>
-    auto is_reachable(
-        const DiGraphBase<Node, Derived>& graph,
-        typename DiGraphBase<Node, Derived>::node_id source,
-        typename DiGraphBase<Node, Derived>::node_id target
-    ) -> bool;
-
-    template <class G>
-    class default_visitor
+    // TODO C++20 better to return a range since this search cannot be interupted from the
+    // visitor
+    // TODO should let user implement reverse with a reverse view when available
+    template <typename Graph, typename Visitor>
+    void
+    dfs_raw(const Graph& graph, Visitor&& visitor, typename Graph::node_id start, bool reverse = false);
+
+    template <typename Graph, typename Visitor>
+    void dfs_raw(const Graph& graph, Visitor&& visitor, bool reverse = false);
+
+    template <typename Graph, typename UnaryFunc>
+    void dfs_preorder_nodes_for_each_id(
```

Comment:
NItpicking: depth_first_search was more expressive than `dfs`. I nunderstand you changed it to avoid very long names for these overloads, can we just omit the `for_each_id` part? So we would have `depth_first_search_preorder_nodes`.

Another (maybe more elegant) solution could be to have all the DFS related stuff into a `depth_first_search` namespace, and remove the `dfs` prefix.

Reply:
I changed to `dfs_` only to mirror `networkx`.

Namespace-wise, this is not even in a graph namespace, only in `mamba::util`. We could improve that as we make a proper utility lib.

Reply:
The idea with all the `for_each` function I write (here and in other places) is to not force a return type like `std::vector` because the user may not need it.

A better API would be in C++20 when, instead of having a visitor pattern, we will return a range that can properly iterate through the desired value without allocating and copying data. At that point we would drop the `for_each`

Reply:
ok, let's keep the `dfs` prefix for now then.